### PR TITLE
Ansible: Add recovery module to recover failed nodes and build and load images locally.

### PR DESCRIPTION
**Pull request by gargpalak048**
# Problem Statement
- How to recover failed nodes
- How to share images locally with the worker nodes

# Design
-  For Bug, Describe the fix here.
-  [POC] Added a recovery module to recover failed nodes
-  [POC] Build and load images locally

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
- [ ] Unit and System Tests are added
- [ ] Test Cases cover Happy Path, Non-Happy Path and Scalability
- [ ] Testing was performed with RPM

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
- [ ] Interface change (if any) are documented
- [ ] Side effects on other features (deployment/upgrade)
- [ ] Dependencies on other component(s)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Documentation
  Checklist for Author
- [ ] Changes done to WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

**Review by Pradeep-B-Thangaraj — APPROVED** [review]
LGTM

**Comment by gargpalak048** [comment]
Test results 

>>> python3 driver.py --playbook playbook-demo.yml --ini /root/inventory

PLAY [all] ***************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [ssc-vm-rhev4-2861.colo.seagate.com]
ok: [ssc-vm-rhev4-2862.colo.seagate.com]

TASK [Check ALL hosts are reacheable] ************************************************************************************
skipping: [ssc-vm-rhev4-2862.colo.seagate.com] => (item=ssc-vm-rhev4-2862.colo.seagate.com)
skipping: [ssc-vm-rhev4-2861.colo.seagate.com] => (item=ssc-vm-rhev4-2862.colo.seagate.com)
skipping: [ssc-vm-rhev4-2862.colo.seagate.com] => (item=ssc-vm-rhev4-2861.colo.seagate.com)
skipping: [ssc-vm-rhev4-2861.colo.seagate.com] => (item=ssc-vm-rhev4-2861.colo.seagate.com)

PLAY [localhost] *********************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [localhost]

TASK [Cleanup of temp files, if present] *********************************************************************************
changed: [localhost] => (item=/tmp/response)
changed: [localhost] => (item=/tmp/failed_nodes_ini)

PLAY [localhost] *********************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [localhost]

TASK [create temp files on remote machine] *******************************************************************************
changed: [localhost] => (item=/tmp/response)
changed: [localhost] => (item=/tmp/failed_nodes_ini)

PLAY [all] ***************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [ssc-vm-rhev4-2861.colo.seagate.com]
ok: [ssc-vm-rhev4-2862.colo.seagate.com]

TASK [Copy and Execute the script] ***************************************************************************************
fatal: [ssc-vm-rhev4-2861.colo.seagate.com]: FAILED! => {"changed": true, "msg": "non-zero return code", "rc": 127, "stderr": "Shared connection to ssc-vm-rhev4-2861.colo.seagate.com closed.\r\n", "stderr_lines": ["Shared connection to ssc-vm-rhev4-2861.colo.seagate.com closed."], "stdout": "/root/.ansible/tmp/ansible-tmp-1658990538.52-118772-240089032152127/cmd.sh: line 1: tree: command not found\r\n", "stdout_lines": ["/root/.ansible/tmp/ansible-tmp-1658990538.52-118772-240089032152127/cmd.sh: line 1: tree: command not found"]}
...ignoring
fatal: [ssc-vm-rhev4-2862.colo.seagate.com]: FAILED! => {"changed": true, "msg": "non-zero return code", "rc": 127, "stderr": "Shared connection to ssc-vm-rhev4-2862.colo.seagate.com closed.\r\n", "stderr_lines": ["Shared connection to ssc-vm-rhev4-2862.colo.seagate.com closed."], "stdout": "/root/.ansible/tmp/ansible-tmp-1658990538.49-118770-104096143575797/cmd.sh: line 1: tree: command not found\r\n", "stdout_lines": ["/root/.ansible/tmp/ansible-tmp-1658990538.49-118770-104096143575797/cmd.sh: line 1: tree: command not found"]}
...ignoring

TASK [Write a json response file] ****************************************************************************************
[WARNING]: The value {'host': 'ssc-vm-rhev4-2861.colo.seagate.com', 'rc': 127, 'stdout': ['/root/.ansible/tmp/ansible-
tmp-1658990538.52-118772-240089032152127/cmd.sh: line 1: tree: command not found']} (type dict) in a string field was
converted to u"{'host': 'ssc-vm-rhev4-2861.colo.seagate.com', 'rc': 127, 'stdout': ['/root/.ansible/tmp/ansible-
tmp-1658990538.52-118772-240089032152127/cmd.sh: line 1: tree: command not found']}" (type string). If this does not look
like what you expect, quote the entire value to ensure it does not change.
changed: [ssc-vm-rhev4-2861.colo.seagate.com -> localhost]
[WARNING]: The value {'host': 'ssc-vm-rhev4-2862.colo.seagate.com', 'rc': 127, 'stdout': ['/root/.ansible/tmp/ansible-
tmp-1658990538.49-118770-104096143575797/cmd.sh: line 1: tree: command not found']} (type dict) in a string field was
converted to u"{'host': 'ssc-vm-rhev4-2862.colo.seagate.com', 'rc': 127, 'stdout': ['/root/.ansible/tmp/ansible-
tmp-1658990538.49-118770-104096143575797/cmd.sh: line 1: tree: command not found']}" (type string). If this does not look
like what you expect, quote the entire value to ensure it does not change.
changed: [ssc-vm-rhev4-2862.colo.seagate.com -> localhost]

PLAY [all] ***************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [ssc-vm-rhev4-2861.colo.seagate.com]
ok: [ssc-vm-rhev4-2862.colo.seagate.com]

TASK [Add entries for failed nodes] **************************************************************************************
changed: [ssc-vm-rhev4-2862.colo.seagate.com -> localhost]
changed: [ssc-vm-rhev4-2861.colo.seagate.com -> localhost]

PLAY RECAP ***************************************************************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
ssc-vm-rhev4-2861.colo.seagate.com : ok=6    changed=3    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1
ssc-vm-rhev4-2862.colo.seagate.com : ok=6    changed=3    unreachable=0    failed=0    skipped=1    rescued=0    ignored=1

========================================
Trying to recover failed Nodes!!!
========================================

PLAY [all] ***************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [ssc-vm-rhev4-2861.colo.seagate.com]
ok: [ssc-vm-rhev4-2862.colo.seagate.com]

TASK [install packages required for project] *****************************************************************************
[DEPRECATION WARNING]: Invoking "yum" only once while using a loop via squash_actions is deprecated. Instead of using a
loop to supply multiple items and specifying `name: "{{ item }}"`, please use `name: ['tree']` and remove the loop. This
feature will be removed in version 2.11. Deprecation warnings can be disabled by setting deprecation_warnings=False in
ansible.cfg.
[DEPRECATION WARNING]: Invoking "yum" only once while using a loop via squash_actions is deprecated. Instead of using a
loop to supply multiple items and specifying `name: "{{ item }}"`, please use `name: ['tree']` and remove the loop. This
feature will be removed in version 2.11. Deprecation warnings can be disabled by setting deprecation_warnings=False in
ansible.cfg.
changed: [ssc-vm-rhev4-2861.colo.seagate.com] => (item=[u'tree'])
changed: [ssc-vm-rhev4-2862.colo.seagate.com] => (item=[u'tree'])

PLAY RECAP ***************************************************************************************************************
ssc-vm-rhev4-2861.colo.seagate.com : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
ssc-vm-rhev4-2862.colo.seagate.com : ok=2    changed=1    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0


PLAY [all] ***************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [ssc-vm-rhev4-2861.colo.seagate.com]
ok: [ssc-vm-rhev4-2862.colo.seagate.com]

TASK [install packages required for project] *****************************************************************************
[DEPRECATION WARNING]: Invoking "yum" only once while using a loop via squash_actions is deprecated. Instead of using a
loop to supply multiple items and specifying `name: "{{ item }}"`, please use `name: ['tree']` and remove the loop. This
feature will be removed in version 2.11. Deprecation warnings can be disabled by setting deprecation_warnings=False in
ansible.cfg.
[DEPRECATION WARNING]: Invoking "yum" only once while using a loop via squash_actions is deprecated. Instead of using a
loop to supply multiple items and specifying `name: "{{ item }}"`, please use `name: ['tree']` and remove the loop. This
feature will be removed in version 2.11. Deprecation warnings can be disabled by setting deprecation_warnings=False in
ansible.cfg.
ok: [ssc-vm-rhev4-2861.colo.seagate.com] => (item=[u'tree'])
ok: [ssc-vm-rhev4-2862.colo.seagate.com] => (item=[u'tree'])

PLAY RECAP ***************************************************************************************************************
ssc-vm-rhev4-2861.colo.seagate.com : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
ssc-vm-rhev4-2862.colo.seagate.com : ok=2    changed=0    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0


PLAY [all] ***************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [ssc-vm-rhev4-2861.colo.seagate.com]
ok: [ssc-vm-rhev4-2862.colo.seagate.com]

TASK [Check ALL hosts are reacheable] ************************************************************************************
skipping: [ssc-vm-rhev4-2862.colo.seagate.com] => (item=ssc-vm-rhev4-2862.colo.seagate.com)
skipping: [ssc-vm-rhev4-2862.colo.seagate.com] => (item=ssc-vm-rhev4-2861.colo.seagate.com)
skipping: [ssc-vm-rhev4-2861.colo.seagate.com] => (item=ssc-vm-rhev4-2862.colo.seagate.com)
skipping: [ssc-vm-rhev4-2861.colo.seagate.com] => (item=ssc-vm-rhev4-2861.colo.seagate.com)

PLAY [localhost] *********************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [localhost]

TASK [Cleanup of temp files, if present] *********************************************************************************
changed: [localhost] => (item=/tmp/response)
changed: [localhost] => (item=/tmp/failed_nodes_ini)

PLAY [localhost] *********************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [localhost]

TASK [create temp files on remote machine] *******************************************************************************
changed: [localhost] => (item=/tmp/response)
changed: [localhost] => (item=/tmp/failed_nodes_ini)

PLAY [all] ***************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [ssc-vm-rhev4-2861.colo.seagate.com]
ok: [ssc-vm-rhev4-2862.colo.seagate.com]

TASK [Copy and Execute the script] ***************************************************************************************
changed: [ssc-vm-rhev4-2862.colo.seagate.com]
changed: [ssc-vm-rhev4-2861.colo.seagate.com]

TASK [Write a json response file] ****************************************************************************************
[WARNING]: The value {'host': 'ssc-vm-rhev4-2861.colo.seagate.com', 'rc': 0, 'stdout': ['.',
'\xe2\x94\x9c\xe2\x94\x80\xe2\x94\x80 dummy.txt', '\xe2\x94\x9c\xe2\x94\x80\xe2\x94\x80 dumy.txt',
'\xe2\x94\x94\xe2\x94\x80\xe2\x94\x80 experiments', '    \xe2\x94\x94\xe2\x94\x80\xe2\x94\x80 reload_network.yml', '', '1
directory, 3 files']} (type dict) in a string field was converted to u"{'host': 'ssc-vm-rhev4-2861.colo.seagate.com',
'rc': 0, 'stdout': ['.', '\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\x80 dummy.txt',
'\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\x80 dumy.txt', '\\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 experiments', '
\\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 reload_network.yml', '', '1 directory, 3 files']}" (type string). If this
does not look like what you expect, quote the entire value to ensure it does not change.
changed: [ssc-vm-rhev4-2861.colo.seagate.com -> localhost]
[WARNING]: The value {'host': 'ssc-vm-rhev4-2862.colo.seagate.com', 'rc': 0, 'stdout': ['.',
'\xe2\x94\x9c\xe2\x94\x80\xe2\x94\x80 docker_container.yml', '\xe2\x94\x9c\xe2\x94\x80\xe2\x94\x80 dummy.txt',
'\xe2\x94\x94\xe2\x94\x80\xe2\x94\x80 dumy.txt', '', '0 directories, 3 files']} (type dict) in a string field was
converted to u"{'host': 'ssc-vm-rhev4-2862.colo.seagate.com', 'rc': 0, 'stdout': ['.',
'\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\x80 docker_container.yml', '\\xe2\\x94\\x9c\\xe2\\x94\\x80\\xe2\\x94\\x80
dummy.txt', '\\xe2\\x94\\x94\\xe2\\x94\\x80\\xe2\\x94\\x80 dumy.txt', '', '0 directories, 3 files']}" (type string). If
this does not look like what you expect, quote the entire value to ensure it does not change.
changed: [ssc-vm-rhev4-2862.colo.seagate.com -> localhost]

PLAY [all] ***************************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [ssc-vm-rhev4-2861.colo.seagate.com]
ok: [ssc-vm-rhev4-2862.colo.seagate.com]

TASK [Add entries for failed nodes] **************************************************************************************
skipping: [ssc-vm-rhev4-2862.colo.seagate.com]
skipping: [ssc-vm-rhev4-2861.colo.seagate.com]

PLAY RECAP ***************************************************************************************************************
localhost                  : ok=4    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
ssc-vm-rhev4-2861.colo.seagate.com : ok=5    changed=2    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0
ssc-vm-rhev4-2862.colo.seagate.com : ok=5    changed=2    unreachable=0    failed=0    skipped=2    rescued=0    ignored=0

**Comment by gargpalak048** [comment]
To build the images locally and save it as a tarball, 
>>> ansible-playbook build.yml --extra-vars "path=/opt/Halo-provisioner/build_and_load_images/build_images tag=cortx-halo:v1"


PLAY [localhost] *********************************************************************************************************

TASK [Gathering Facts] ***************************************************************************************************
ok: [localhost]

TASK [build a halo container image] **************************************************************************************
changed: [localhost]

TASK [save halo container image as a tarball] ****************************************************************************
changed: [localhost]

PLAY RECAP ***************************************************************************************************************
localhost                  : ok=3    changed=2    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0

>>> docker images
REPOSITORY   TAG       IMAGE ID       CREATED              SIZE
cortx-halo   v1        eb90be27b037   About a minute ago   920MB

**Review by shaileshsvc — APPROVED** [review]

File: Halo-prvsnr/ansible/build_and_load_images/build_images/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,6 @@
+FROM python:3.10
```
Comment (ajaysrivas) [inline Halo-prvsnr/ansible/build_and_load_images/build_images/Dockerfile:1]: This repo hosts code written for POCs and if the poc is specific to the technology then it should go to corresponding technology folder. Please change base directory to ansible.

**Review by ajaysrivas — CHANGES_REQUESTED** [review]

File: Halo-prvsnr/ansible/build_and_load_images/build_images/Dockerfile (hunk 1 above)
Comment (gargpalak048) [reply Halo-prvsnr/ansible/build_and_load_images/build_images/Dockerfile]: Ack.
change the base dir from Halo-prvsnr/ansible to ansible

**Review by ajaysrivas — APPROVED** [review]

**Pull request merged**